[CK_BUILDER] Refactor builder factory code.

shumway opened the pull request:
This PR refactors the builder factory code subdirectories and a `ck_tile::builder::factory namespace`.

The factory implements compile-time dispatch from high-level signature and algorithm descriptors to our exisiting specialized convolution kernel implementations.

Major changes in this PR:

1. Dispatch logic is explicit in the function `make_conv_instance` instead of implicit in template specialization selection.
2. Helper code is moved to a subdirectory `builder/factory/helpers`.
3. Helpers now have unit tests.
4. Factories are moved to their own files.
5. Code moved to namespaces `ck_tile::builder::factory` and `ck_tile::builder::factory::internal`.

This does not yet fix the problem of bad error messages, but the `make_conv_instance` function makes the poor error messages clear. The choice of algorithm must be much more robust (perhaps with explicit enumeration in the algorithm descriptor), so that the dispatch doesn't fail.

Quality changes:
* Making dispatch explicit rather than implicit will improve robustness, readability, maintainability, testability, and extensibility.
* Separating code into separate files and subdirectories helps readability and extensibility.
* Adding unit tests for helpers documents behavior and will enable more complex logic and functionality.
* Separating files (especially unit tests) helps clarify includes and dependencies and makes code easier to refactor.

## Design Overview

The factory system operates in two phases:

1. **Algorithm Classification**: The function `make_conv_instance` in  `conv_dispatcher.hpp` inspects the signature and algorithm descriptors to determine which kernel variant they satisfy (XDL V3, XDL, WMMA, DL, or Large Tensor)

2. **Factory Instantiation**: Each factory (`conv_fwd_*_factory.hpp`) transforms builder descriptors into CK device operation template parameters and instantiates the corresponding kernel device operation.

## Key Files

- **`conv_dispatcher.hpp`**: Entry point with `make_conv_instance()` function. Contains dispatch logic and algorithm classification predicates. **Start here** to understand the overall flow.

- **`conv_fwd_*_factory.hpp`**: Individual factories for each kernel variant. Each extracts configuration from descriptors, validates parameters, and instantiates the underlying CK device operation.

- **`helpers/`**: Transformation utilities that map builder types to CK device operation parameters (layouts, data types, elementwise ops, block configurations, etc.)

## Usage

```cpp
#include "ck_tile/builder/factory/conv_dispatcher.hpp"

using Factory = decltype(make_conv_instance<signature, algorithm, "v1">());
```

The dispatcher automatically selects the appropriate factory following explicit logic.

shumway committed:
Split conv_factory.hpp into separate files

shumway committed:
Convert to dispatching through a function

The template specialization and resolution was not working. In particular, SFINAE caused all the helpful errors from asserts to be hidden. We make the dispatch explicit through a function and get rid of the template specialization.

shumway committed:
Update builder factory namespaces

* `ck_tile::builder::factory` for factories.
* `ck_tile::builder::factory::interal` for factory helpers.

shumway committed:
Clean up convolution dispatcher

The concepts for each factory implemenation are literally implementation details rather than domain concepts. Move that logic into helper functions to make it easier to study and refactor.

shumway committed:
Add unit tests for factory helpers

shumway committed:
Add README.md file for the factory subdirectory

shumway committed:
Change paramters in test_conv_discription to fix gfx950

We're seeing a build error with this test on gfx950, so we make a cosmetic change to the parameters in this description test. It's better to do a quick fix so we can get this under CI.

Snektron commented on experimental/builder/include/ck_tile/builder/factory/conv_dispatcher.hpp:
```diff
@@ -0,0 +1,191 @@
+
+// XDL-based kernel with V3 pipeline structure (newer block GEMM pipeline)
+template <typename T>
+consteval bool IsXdlV3Algorithm()
+{
+    return ConvAlgorithmDescriptor<T> && SpecifiesThreadBlock<T> && SpecifiesGridwiseXdlGemm<T> &&
+           SpecifiesBlockTransfer<T> && SpecifiesLdsTransfer<T> &&
+           SpecifiesThreadClusterAccessOrder<T> && SpecifiesSourceAccessOrder<T> &&
+           SpecifiesFwdConcSpecialization<T> && SpecifiesGemmSpecialization<T> &&
+           SpecifiesBlockGemm<T>;
+}
```
You can make this a concept. It can be used in an if-condition too, except that you don't need to call it like a function. Just `if constexpr(IsXdlV3Algorithm<...>)`.

shumway replied:
It was written as a concept, and the PR changes it to not be a concept. The reason is that we need to be very explicit about the logic for selecting templates now. Also, even though this has the functionality of a concept, it is an implementation detail. Concepts are semantic categories. I'm leaving this code in this verbose state to set design changes. I'll add a TODO and some comments to make this more clear.

Snektron commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_elementwise_op.hpp:
```diff
@@ -0,0 +1,37 @@
+};
+
+template <>
+struct ElementwiseOps<ElementwiseOperation::PASS_THROUGH>
+{
+    using AElementwiseOp   = ck::tensor_operation::element_wise::PassThrough;
+    using BElementwiseOp   = ck::tensor_operation::element_wise::PassThrough;
+    using CDEElementwiseOp = ck::tensor_operation::element_wise::PassThrough;
+};
```
At some point we will have to make this struct work with CK Tile also. For now its fine, just a heads up that I think that this is not going to be the final fork.

shumway replied:
This is why I'm refactoring out the helpers. They are going to get a lot more refactoring and extension as we add CK Tile and backwards kernels. Likely we'll have some automatic calculation and defaults here, too. Isolating this logic and setting up testing will help us evolve this quickly.

vpietila-amd commented on experimental/builder/include/ck_tile/builder/factory/conv_dispatcher.hpp:
```diff
@@ -0,0 +1,191 @@
+constexpr auto make_conv_instance()
+{
+    if constexpr(ConvDirectionIsForward<SIGNATURE>)
+    {
```
Perhaps we could define template specializations for each convolution direction to avoid having so many `if-else` branches?

vpietila-amd commented:
The [Readme file](https://github.com/ROCm/composable_kernel/blob/jshumway/dispatch/experimental/builder/README.md) at the the root of the CK Builder directory has section "Directory Structure". That should be updated to reflect the new organization of the builder code. I would also link the [Builder Factory Readme](https://github.com/ROCm/composable_kernel/blob/jshumway/dispatch/experimental/builder/include/ck_tile/builder/factory/README.md) to the "master" Readme so that we have one entrypoint to the builder documentation.

shumway committed:
Update README.md and comments  for dispatch refactor

aosewski commented on experimental/builder/include/ck_tile/builder/factory/conv_dispatcher.hpp:
```diff
@@ -0,0 +1,196 @@
+// - **WMMA**: Wavefront Matrix-Matrix Accumulate variant optimized for WMMA-capable hardware.
+//   Requires similar configuration to XDL.
+//
+// - **DL**: Specialized Direct Load kernel optimized for specific data layouts (NHWC/KYXC/NHWK).
```
This is an error. The `DL` is not a direct load, But in very old CK kernels targeted for RDNA architectures there was just such naming convention to make them different from XDL (MFMA) kernels. 
What this acronym stands for? I'm not sure.. Deep Learning ? :P

shumway replied:
LOL, yes, the initial pass of this documentation is AI, since I haven't dug into the kernel implementations yet. I'll fix and iterate with you.

shumway replied:
I updated this to not hallucinate "direct load" or "dot product and load" or "domain language" but rather "deep learning".

aosewski commented on experimental/builder/test/unit_conv_elementwise_op.cpp:
```diff
@@ -0,0 +1,38 @@
+
+namespace {
+
+// namespace ckb = ::ck_tile::builder;
```
Leftover?

shumway replied:
removed

aosewski commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_thread_block.hpp:
```diff
@@ -0,0 +1,32 @@
+{
+    size_t block_size = 0;
+    MNK per_block     = {};
+};
```
Honestly speaking I don't much like this representation. I prefer to have sth like this: https://github.com/ROCm/composable_kernel/blob/develop/experimental/builder/include/ck_tile/builder/reflect/conv_traits.hpp#L116-L121 which is IMHO more understandable and more related to problem context. Although I understand that this adheres more to the implementation.

shumway replied:
I fixed this partway. This is a big refactoring PR, so I don't want to make too many extra changes that might break logic. I'll collect all these things, look at them all critically, and address in a follow-up PR.

aosewski commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_block_transfer.hpp:
```diff
@@ -0,0 +1,71 @@
+constexpr BlockTransfer SetFwdConvBlockTransfer()
+{
+    constexpr auto& TCL = TRANSFER.block_transfer;
+    constexpr auto& TCO = TRANSFER.block_transfer_access_order;
```
It took me a moment to find out what those acronyms stands for...

shumway replied:
I renamed these aliases `block_xfer`, `block_order`, `src_order`, and `lds_config`.

aosewski commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_block_transfer.hpp:
```diff
@@ -0,0 +1,71 @@
+struct CBlockTransfer
+{
+    size_t m_per_wave_per_shuffle            = 0;
+    size_t n_per_wave_per_shuffle            = 0;
```
Those names are confusing. Originally this is `CShuffleMXdlPerWavePerShuffl` or `CShuffleNXdlPerWavePerShuffl` and stands for https://github.com/ROCm/composable_kernel/blob/develop/experimental/builder/include/ck_tile/builder/reflect/conv_traits.hpp#L184-L190 I'd rename it to sth like `m_xdl_....` and `n_xdl...` to make it clear.

shumway replied:
These changes are getting risky in a big refactoring PR. I'm going to collect these issues and address them in a follow-up PR.

aosewski commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_block_transfer.hpp:
```diff
@@ -0,0 +1,71 @@
+                                  .thread_cluster_dims =
+                                      {
+                                          TCL.m_block,
+                                          TCL.m_wave_per_xdl,
```
This name is confusing and misleading. Originally those lengths are: `MBlock_MPerBlock_NBlock_NPerBlock`

shumway replied:
These changes are getting risky in a refactoring PR. I will go through and address them in a followup PR. I don't want to accidentally change logic.

aosewski commented on experimental/builder/include/ck_tile/builder/factory/helpers/conv_block_transfer.hpp:
```diff
@@ -0,0 +1,71 @@
+                                          TCL.m_block,
+                                          TCL.m_wave_per_xdl,
+                                          TCL.n_block,
+                                          TCL.n_wave_per_xdl,
```
As above.

shumway replied:
done

shumway commented:
> The [Readme file](https://github.com/ROCm/composable_kernel/blob/jshumway/dispatch/experimental/builder/README.md) at the the root of the CK Builder directory has section "Directory Structure". That should be updated to reflect the new organization of the builder code. I would also link the [Builder Factory Readme](https://github.com/ROCm/composable_kernel/blob/jshumway/dispatch/experimental/builder/include/ck_tile/builder/factory/README.md) to the "master" Readme so that we have one entrypoint to the builder documentation.

Done.

shumway committed:
Update experimental/builder/README.md

Clarify integration tests description from review comment.

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

shumway committed:
Address reviewer comments.

Cleanup names and comments based on reviewer feedback.

shumway committed:
Fix failing tests and build failures

We had suggested changes during code review. In hindsight, we should not have included those during a large refactor and put them in a following cleanup PR. These reviews and code merges left the code in a broken state, so this commit has some small changes to fix tests and fix the build.

shumway commented:
We need to get this merged. I should have rejected the suggested changes because this is a refactoring PR. I accepted some of them, but that left the code in a broken state. I just patched it up and it should pass CI now. There is a lot more cleanup and alignment of notation and design needed, and this PR is critical to split apart the code and make that all easier to change and review.
